runtime/tr.source/trj9/z/codegen/ReduceSynchronizedFieldLoad.cpp

```diff
@@ -282,292 +265,191 @@ ReduceSynchronizedFieldLoad::performOnTreeTops(TR::TreeTop* startTreeTop, TR::Tr
+            break;
             }
          }
       }
 
    return transformed;
    }
 
-bool
-ReduceSynchronizedFieldLoad::advanceIterator(TR::TreeTopIterator& iter, TR::TreeTop* endTreeTop)
+TR::Node*
+ReduceSynchronizedFieldLoad::findLoadInSynchornizedRegion(TR::TreeTop* startTreeTop, TR::TreeTop* endTreeTop, TR::TreeTop* monentTreeTop, TR::TreeTop* monexitTreeTop, TR::Node* synchronizedObjectNode)
    {
-   ++iter;
-
-   if (iter == endTreeTop)
+   TR::PreorderNodeIterator iter(startTreeTop, cg->comp());
+
+   // First iterate though all the nodes from the start treetop until we reach the monitor provided so that all nodes
+   // seen thus far would have already been visited, and hence we will not recurse into them in the subsequent for loop
+   // since a reference was already seen. This enables us to carry out the reduce synchronized field load optimization
+   // even if there are sideeffect nodes within the monitored region - as long as those sideeffect nodes have been
+   // evaluated outside of the monitored region.
+   for (; iter != monentTreeTop->getNextTreeTop(); ++iter)
       {
-      --iter;
-
-      TR::DebugCounter::incStaticDebugCounter(cg->comp(), TR::DebugCounter::debugCounterName(cg->comp(), "codegen/z/ReduceSynchronizedFieldLoad/failure/advanceIterator/%s", cg->comp()->signature()));
+      TR::Node* currentNode = iter.currentNode();
 
-      return false;
+      if (cg->comp()->getOption(TR_TraceCG))
+         {
+         traceMsg(cg->comp(), "Iterating node [%p] outside the monitored region\n", currentNode);
+         }
       }
 
-   if (cg->comp()->getOption(TR_TraceCG))
-      {
-      traceMsg(cg->comp(), "Advancing iterator to [%p]\n", iter.currentNode());
-      }
+   TR::Node* loadNode = NULL;
 
-   // Synchronized region may span multiple blocks
-   if (iter.currentNode()->getOpCodeValue() == TR::BBEnd)
+   for (; iter != monexitTreeTop; ++iter)
       {
-      TR::TreeTop* nextTreeTop = iter.currentTree()->getNextTreeTop();
+      TR::Node* currentNode = iter.currentNode();
 
-      if (nextTreeTop == NULL || nextTreeTop == endTreeTop || !nextTreeTop->getNode()->getBlock()->isExtensionOfPreviousBlock())
+      if (cg->comp()->getOption(TR_TraceCG))
          {
-         TR::DebugCounter::incStaticDebugCounter(cg->comp(), TR::DebugCounter::debugCounterName(cg->comp(), "codegen/z/ReduceSynchronizedFieldLoad/failure/advanceIterator/%s", cg->comp()->signature()));
-
-         return false;
+         traceMsg(cg->comp(), "Iterating node [%p] inside the monitored region\n", currentNode);
          }
-      else
-         {
-         ++iter;
 
-         // Sanity check that a new block starts after the previous
-         TR_ASSERT(iter.currentNode()->getOpCodeValue() == TR::BBStart, "Expecting a BBStart treetop but found %s\n", iter.currentNode()->getOpCode().getName());
+      TR::ILOpCode opcode = currentNode->getOpCode();
 
-         ++iter;
+      if (opcode.hasSymbolReference() || opcode.isBranch())
```

Comment:
I get that use of this function is to find a load node within synchronized region. Let me know if I am wrong here, but in the synchronized region, if a node is branch node or it has a symbol reference but is not a candidate for synchronized field load, for these kind of side effect nodes, this optimization will be disabled, right?

Reply:
@r30shah yes that is correct. The only exit condition out of the for loop is the `break` statement towards the end (other than reaching the `monexit` that is). If you follow the logic to reach the `break` it reads in English as follows:

If there exists a node in the synchronized region that has a symbol reference or is a branch then unless this node is an indirect load off the synchronized object and no previous such load has been seen then terminate the for loop and return that no candidate has been found. Otherwise return the single load (if any) as the candidate.

Reply:
@fjeremic Thanks for the explanation. Changes looks good to me.